### ISSUE TITLE
subtract a cube from polyhedron freezes in "processing, please wait"

### ISSUE THREAD

**Issue by hrgdavor** [jscad/OpenJSCAD.org#754]
## Expected Behavior
cube subtracted from a pyramid

## Actual Behavior
render hangs

## Steps to Reproduce the Problem

```
const jscad = require('@jscad/modeling')

function main(){
  var a = 20, height = 28
  var poly =  jscad.primitives.polyhedron({
    points:[
      [-a,-a],
      [ a,-a],
      [ a, a],
      [-a, a],
      [0,0,height+1],
    ],
    faces:[
      [3,2,1,0],
      [0,1,4],
      [1,2,4],
      [2,3,4],
      [3,0,4],
    ]
  })

  var cube = jscad.primitives.cuboid({size:[3,3,3]})
  
  // uncomment to view
  // return [cube, poly] // preview

  return  jscad.booleans.subtract(
  	poly,
  	jscad.primitives.cuboid({size:[3,3,3]}),
  )
}

module.exports = {main}

```

![image](https://user-images.githubusercontent.com/2480762/103797674-1334e380-5049-11eb-86c6-f50e2aafa32d.png)

## Specifications

  - Version: V2
  - Platform:
  - Environment: (browser)

**Comment by z3dev** [comment]
@hrgdavor welcome again! Just a quick question, is the list of points correct in the example? Some of the points are not 3D points.

 ```
   points:[
      [-a,-a],  // here
      [ a,-a],
      [ a, a],
      [-a, a],
      [0,0,height+1],
    ],
```

**Comment by hrgdavor** [comment]
:) it works when I add z=0

```
    points:[
      [-a,-a, 0],
      [ a,-a, 0],
      [ a, a, 0],
      [-a, a, 0],
      [0,0,height+1],
    ],

```

I was confused because the object was shown without boolean operations, so I assumed the code was fine when I ommit `z` 

Although it is pretty obvious now, I did not think to test with `z` ... it was code from my other script in jscad V1,
and even then I did not do boolean on it so if the same issue was present then I did not run into it :)

thanks for the quick response.

I would like to leave the issue open for now, because I am guessing you would either fix `z` or throw error instead of this behavior.

polyhedron is a pretty neat way to produce 3D objects and now that I am starting to understand it a bit, I have ideas ...

**Comment by z3dev** [comment]
@hrgdavor thanks for the quick feedback.

Ok. So, should there be checks performed on the points provided? Should checks be performed on the polygons created?

**Comment by hrgdavor** [comment]
Well, I am not the one to define standards four your lib :), but here is my opinion based on what I have seen while using jscad.

Some other code I called generated an error when I provided wrong type of data, I would prefer that polyhedron throws an error if point is not [x,y,z] and also an error for faces that reference a non existent point.

currently if I provide a non existent point in faces section it shows an error down the line:
"Cannot read property '0' of undefined" 

... which is not useful unless you debug ... 
I would be better to say "point  number 5 is not defined"  ... this would be useful and I would immediately know I have a bad reference in the faces section.
